File: src/conformance.cpp
#include <spdlog/spdlog.h>
#include <chrono>
#include <filesystem>
#include <fstream>
#include <sstream>

#include "dcpdoctor/conformance.h"
#include "dcpdoctor/compliance.h"
#include "dcpdoctor/cpl.h"
#include "dcpdoctor/validators.h"

namespace fs = std::filesystem;

namespace dcpdoctor
{

static std::string today_iso()
{
  auto now = std::chrono::system_clock::now();
  auto days = std::chrono::floor<std::chrono::days>(now);
  std::chrono::year_month_day ymd{days};
  char buf[16];
  std::snprintf(buf, sizeof(buf), "%04d-%02u-%02u",
                static_cast<int>(ymd.year()),
                static_cast<unsigned>(ymd.month()),
                static_cast<unsigned>(ymd.day()));
  return buf;
}

static ConformanceTest make_test(const std::string& id, const std::string& desc,
                                 const std::string& spec, bool passed,
                                 const std::string& detail = {})
{
  ConformanceTest t;
  t.test_id = id;
  t.description = desc;
  t.spec_reference = spec;
  t.passed = passed;
  t.detail = detail;
  return t;
}

ConformanceReport run_conformance_tests(const ConformanceOptions& opts)
{
  ConformanceReport report;
  report.report_date = today_iso();
  report.tool_version = "dcpdoctor 1.0";
  report.dcp_dir = opts.dcp_dir;

  if(!fs::exists(opts.dcp_dir) || !fs::is_directory(opts.dcp_dir))
  {
    report.error = "DCP directory not found: " + opts.dcp_dir.string();
    return report;
  }

  // --- Structure tests (SMPTE ST 429-9) ---

  bool has_assetmap = fs::exists(opts.dcp_dir / "ASSETMAP") ||
                      fs::exists(opts.dcp_dir / "ASSETMAP.xml");
  report.structure_tests.push_back(
      make_test("DCI-STRUCT-1", "ASSETMAP present", "SMPTE ST 429-9:2014",
                has_assetmap, has_assetmap ? "Found" : "Missing"));

  bool has_volindex = fs::exists(opts.dcp_dir / "VOLINDEX") ||
                      fs::exists(opts.dcp_dir / "VOLINDEX.xml");
  report.structure_tests.push_back(
      make_test("DCI-STRUCT-2", "VOLINDEX present", "SMPTE ST 429-9:2014",
                has_volindex, has_volindex ? "Found" : "Missing"));

  // Find PKL
  std::vector<fs::path> pkls;
  for(const auto& entry : fs::directory_iterator(opts.dcp_dir))
  {
    if(!entry.is_regular_file())
      continue;
    if(entry.path().extension() != ".xml")
      continue;
    std::ifstream f(entry.path());
    std::string line;
    for(int i = 0; i < 5 && std::getline(f, line); ++i)
    {
      if(line.find("PackingList") != std::string::npos)
      {
        pkls.push_back(entry.path());
        break;
      }
    }
  }
  report.structure_tests.push_back(
      make_test("DCI-STRUCT-3", "PackingList (PKL) present", "SMPTE ST 429-8:2014",
                !pkls.empty(), pkls.empty() ? "No PKL found" : "Found " + std::to_string(pkls.size())));

  // Find CPL
  std::vector<fs::path> cpls;
  std::optional<Cpl> first_cpl;
  for(const auto& entry : fs::directory_iterator(opts.dcp_dir))
  {
    if(!entry.is_regular_file())
      continue;
    auto cpl = Cpl::parse(entry.path());
    if(cpl.has_value() && !cpl->id.empty())
    {
      cpls.push_back(entry.path());
      if(!first_cpl.has_value())
        first_cpl = cpl;
    }
  }
  report.structure_tests.push_back(
      make_test("DCI-STRUCT-4", "CompositionPlaylist (CPL) present", "SMPTE ST 429-7:2006",
                !cpls.empty(), cpls.empty() ? "No CPL found" : "Found " + std::to_string(cpls.size())));

  // MXF files
  bool has_mxf = false;
  int mxf_count = 0;
  for(const auto& entry : fs::directory_iterator(opts.dcp_dir))
  {
    if(entry.path().extension() == ".mxf")
    {
      has_mxf = true;
      mxf_count++;
    }
  }
  report.structure_tests.push_back(
      make_test("DCI-STRUCT-5", "MXF track files present", "SMPTE ST 429-3:2006",
                has_mxf, std::to_string(mxf_count) + " MXF file(s)"));

  // --- CPL tests (SMPTE ST 429-7) ---
  if(first_cpl.has_value())
  {
    report.content_title = first_cpl->content_title;
    report.cpl_id = first_cpl->id;
    report.issue_date = first_cpl->issue_date;

    // CPL has valid UUID
    bool valid_id = first_cpl->id.find("urn:uuid:") == 0 &&
                    first_cpl->id.size() > 20;
    report.cpl_tests.push_back(
        make_test("DCI-CPL-1", "CPL Id is valid URN UUID", "SMPTE ST 429-7:2006 §6.1",
                  valid_id, first_cpl->id));

    // CPL has content title
    report.cpl_tests.push_back(
        make_test("DCI-CPL-2", "ContentTitleText present", "SMPTE ST 429-7:2006 §6.2",
                  !first_cpl->content_title.empty(), first_cpl->content_title));

    // CPL has content kind
    report.cpl_tests.push_back(
        make_test("DCI-CPL-3", "ContentKind present", "SMPTE ST 429-7:2006 §6.4",
                  !first_cpl->content_kind.empty(), first_cpl->content_kind));

    // CPL has at least one reel
    report.cpl_tests.push_back(
        make_test("DCI-CPL-4", "At least one Reel present", "SMPTE ST 429-7:2006 §6.10",
                  !first_cpl->reels.empty(),
                  std::to_string(first_cpl->reels.size()) + " reel(s)"));

    // Each reel has picture
    bool all_reels_have_pic = true;
    for(const auto& reel : first_cpl->reels)
    {
      if(reel.picture.id.empty() && reel.picture.duration == 0)
      {
        all_reels_have_pic = false;
        break;
      }
    }
    report.cpl_tests.push_back(
        make_test("DCI-CPL-5", "All reels have MainPicture", "SMPTE ST 429-7:2006 §6.10.1",
                  all_reels_have_pic));

    // Issue date present
    report.cpl_tests.push_back(
        make_test("DCI-CPL-6", "IssueDate present", "SMPTE ST 429-7:2006 §6.3",
                  !first_cpl->issue_date.empty(), first_cpl->issue_date));
  }

  // --- Picture tests (SMPTE ST 429-4) ---
  if(opts.check_picture_profile)
  {
    // Check for valid picture MXFs (at minimum, verify they exist and have size > 0)
    for(const auto& entry : fs::directory_iterator(opts.dcp_dir))
    {
      if(entry.path().extension() != ".mxf")
        continue;
      if(entry.file_size() == 0)
      {
        report.picture_tests.push_back(
            make_test("DCI-PIC-1", "MXF file has non-zero size", "SMPTE ST 429-3:2006",
                      false, entry.path().filename().string() + " is empty"));
      }
    }
    if(report.picture_tests.empty())
    {
      report.picture_tests.push_back(
          make_test("DCI-PIC-1", "All MXF files have non-zero size", "SMPTE ST 429-3:2006",
                    has_mxf));
    }
  }

  // --- Audio tests ---
  // Basic: check that sound assets are referenced
  if(first_cpl.has_value())
  {
    bool has_audio = false;
    for(const auto& reel : first_cpl->reels)
    {
      if(!reel.sound.id.empty())
      {
        has_audio = true;
        break;
      }
    }
    report.audio_tests.push_back(
        make_test("DCI-AUD-1", "Audio track referenced in CPL", "SMPTE ST 429-7:2006",
                  has_audio, has_audio ? "MainSound present" : "No MainSound in any reel"));
  }

  // --- Security tests ---
  if(opts.check_security)
  {
    // Check for encryption indicators
    bool has_encryption = false;
    for(const auto& cpl_path : cpls)
    {
      std::ifstream f(cpl_path);
      std::string content((std::istreambuf_iterator<char>(f)),
                          std::istreambuf_iterator<char>());
      if(content.find("KeyId") != std::string::npos)
        has_encryption = true;
    }
    report.security_tests.push_back(
        make_test("DCI-SEC-1", "Encryption status detected", "SMPTE ST 429-6:2006",
                  true, // informational, not a pass/fail
                  has_encryption ? "Encrypted (KeyId found)" : "Not encrypted"));
  }

  // --- Detect standard ---
  for(const auto& cpl_path : cpls)
  {
    std::ifstream f(cpl_path);
    std::string content((std::istreambuf_iterator<char>(f)),
                        std::istreambuf_iterator<char>());
    if(content.find("smpte-ra.org") != std::string::npos)
      report.detected_standard = Standard::smpte;
    else if(content.find("digicine.com") != std::string::npos ||
            content.find("cinecert.com") != std::string::npos)
      report.detected_standard = Standard::interop;
  }

  // --- Summarize ---
  auto count_tests = [&](const std::vector<ConformanceTest>& tests) {
    for(const auto& t : tests)
    {
      report.total_tests++;
      if(t.passed)
        report.tests_passed++;
      else
        report.tests_failed++;
    }
  };
  count_tests(report.structure_tests);
  count_tests(report.cpl_tests);
  count_tests(report.picture_tests);
  count_tests(report.audio_tests);
  count_tests(report.security_tests);

  report.conformant = (report.tests_failed == 0);
  spdlog::info("DCI conformance: {}/{} tests passed ({})",
               report.tests_passed, report.total_tests,
               report.conformant ? "CONFORMANT" : "NON-CONFORMANT");
  return report;
}

static void json_escape(std::ostream& out, const std::string& s)
{
  for(char c : s)
  {
    switch(c)
    {
      case '"': out << "\\\""; break;
      case '\\': out << "\\\\"; break;
      case '\n': out << "\\n"; break;
      default: out << c;
    }
  }
}

std::string conformance_to_json(const ConformanceReport& report)
{
  std::ostringstream json;
  json << "{\n";
  json << "  \"tool_version\": \"" << report.tool_version << "\",\n";
  json << "  \"report_date\": \"" << report.report_date << "\",\n";
  json << "  \"dcp_dir\": \"";
  json_escape(json, report.dcp_dir.string());
  json << "\",\n";
  json << "  \"content_title\": \"";
  json_escape(json, report.content_title);
  json << "\",\n";
  json << "  \"cpl_id\": \"" << report.cpl_id << "\",\n";
  json << "  \"detected_standard\": \""
       << (report.detected_standard == Standard::smpte ? "SMPTE" :
           report.detected_standard == Standard::interop ? "Interop" : "Unknown")
       << "\",\n";
  json << "  \"conformant\": " << (report.conformant ? "true" : "false") << ",\n";
  json << "  \"total_tests\": " << report.total_tests << ",\n";
  json << "  \"tests_passed\": " << report.tests_passed << ",\n";
  json << "  \"tests_failed\": " << report.tests_failed << ",\n";

  auto write_tests = [&](const std::string& name, const std::vector<ConformanceTest>& tests) {
    json << "  \"" << name << "\": [\n";
    for(size_t i = 0; i < tests.size(); ++i)
    {
      const auto& t = tests[i];
      json << "    {\"id\": \"" << t.test_id << "\", "
           << "\"description\": \"";
      json_escape(json, t.description);
      json << "\", "
           << "\"spec\": \"" << t.spec_reference << "\", "
           << "\"passed\": " << (t.passed ? "true" : "false");
      if(!t.detail.empty())
      {
        json << ", \"detail\": \"";
        json_escape(json, t.detail);
        json << "\"";
      }
      json << "}";
      if(i + 1 < tests.size())
        json << ",";
      json << "\n";
    }
    json << "  ]";
  };

  write_tests("structure_tests", report.structure_tests);
  json << ",\n";
  write_tests("cpl_tests", report.cpl_tests);
  json << ",\n";
  write_tests("picture_tests", report.picture_tests);
  json << ",\n";
  write_tests("audio_tests", report.audio_tests);
  json << ",\n";
  write_tests("security_tests", report.security_tests);
  json << "\n}\n";
  return json.str();
}

std::string conformance_to_html(const ConformanceReport& report)
{
  std::ostringstream html;
  html << "<!DOCTYPE html>\n<html><head><meta charset=\"utf-8\">\n";
  html << "<title>DCI Conformance Report</title>\n";
  html << "<style>\n"
       << "body { font-family: -apple-system, sans-serif; max-width: 900px; margin: 2em auto; }\n"
       << "h1 { border-bottom: 2px solid #333; }\n"
       << "table { border-collapse: collapse; width: 100%; margin: 1em 0; }\n"
       << "th, td { border: 1px solid #ddd; padding: 8px; text-align: left; }\n"
       << "th { background: #f5f5f5; }\n"
       << ".pass { color: #2e7d32; font-weight: bold; }\n"
       << ".fail { color: #c62828; font-weight: bold; }\n"
       << ".badge { display: inline-block; padding: 4px 12px; border-radius: 4px; "
       << "font-weight: bold; color: white; }\n"
       << ".badge-pass { background: #2e7d32; }\n"
       << ".badge-fail { background: #c62828; }\n"
       << "</style>\n</head><body>\n";

  html << "<h1>DCI Conformance Report</h1>\n";
  html << "<p><strong>Tool:</strong> " << report.tool_version << "</p>\n";
  html << "<p><strong>Date:</strong> " << report.report_date << "</p>\n";
  html << "<p><strong>DCP:</strong> " << report.dcp_dir.string() << "</p>\n";
  if(!report.content_title.empty())
    html << "<p><strong>Title:</strong> " << report.content_title << "</p>\n";
  html << "<p><strong>Standard:</strong> "
       << (report.detected_standard == Standard::smpte ? "SMPTE" :
           report.detected_standard == Standard::interop ? "Interop" : "Unknown")
       << "</p>\n";

  html << "<p><span class=\"badge "
       << (report.conformant ? "badge-pass" : "badge-fail") << "\">"
       << (report.conformant ? "CONFORMANT" : "NON-CONFORMANT")
       << "</span> " << report.tests_passed << "/" << report.total_tests
       << " tests passed</p>\n";

  auto write_table = [&](const std::string& title, const std::vector<ConformanceTest>& tests) {
    if(tests.empty())
      return;
    html << "<h2>" << title << "</h2>\n";
    html << "<table><tr><th>ID</th><th>Test</th><th>Spec</th><th>Result</th><th>Detail</th></tr>\n";
    for(const auto& t : tests)
    {
      html << "<tr><td>" << t.test_id << "</td><td>" << t.description
           << "</td><td>" << t.spec_reference << "</td><td class=\""
           << (t.passed ? "pass" : "fail") << "\">"
           << (t.passed ? "PASS" : "FAIL") << "</td><td>" << t.detail
           << "</td></tr>\n";
    }
    html << "</table>\n";
  };

  write_table("Structure", report.structure_tests);
  write_table("Composition Playlist", report.cpl_tests);
  write_table("Picture", report.picture_tests);
  write_table("Audio", report.audio_tests);
  write_table("Security", report.security_tests);

  html << "</body></html>\n";
  return html.str();
}

} // namespace dcpdoctor

File: rust/crates/dcpdoctor-core/src/conformance.rs
//! DCI conformance test suite — structured pass/fail reporting.

use std::path::{Path, PathBuf};

use serde::Serialize;

use crate::Standard;

/// A single conformance test result.
#[derive(Debug, Clone, Serialize)]
pub struct ConformanceTest {
    pub test_id: String,
    pub description: String,
    pub spec_reference: String,
    pub passed: bool,
    pub detail: String,
}

/// Options for conformance testing.
pub struct ConformanceOptions {
    pub dcp_dir: PathBuf,
    pub check_picture_profile: bool,
    pub check_security: bool,
}

/// Full conformance report.
#[derive(Debug, Clone, Default, Serialize)]
pub struct ConformanceReport {
    pub tool_version: String,
    pub report_date: String,
    pub dcp_dir: PathBuf,
    pub content_title: String,
    pub cpl_id: String,
    pub issue_date: String,
    pub detected_standard: Standard,
    pub conformant: bool,
    pub total_tests: u32,
    pub tests_passed: u32,
    pub tests_failed: u32,
    pub error: String,
    pub structure_tests: Vec<ConformanceTest>,
    pub cpl_tests: Vec<ConformanceTest>,
    pub picture_tests: Vec<ConformanceTest>,
    pub audio_tests: Vec<ConformanceTest>,
    pub security_tests: Vec<ConformanceTest>,
}

/// Run DCI conformance tests on a DCP.
pub fn run_conformance_tests(opts: &ConformanceOptions) -> ConformanceReport {
    let mut report = ConformanceReport {
        tool_version: "dcpdoctor 1.0".into(),
        report_date: today_iso(),
        dcp_dir: opts.dcp_dir.clone(),
        ..Default::default()
    };

    if !opts.dcp_dir.exists() || !opts.dcp_dir.is_dir() {
        report.error = format!("DCP directory not found: {}", opts.dcp_dir.display());
        return report;
    }

    // --- Structure tests (SMPTE ST 429-9) ---
    let has_assetmap =
        opts.dcp_dir.join("ASSETMAP").exists() || opts.dcp_dir.join("ASSETMAP.xml").exists();
    report.structure_tests.push(make_test(
        "DCI-STRUCT-1",
        "ASSETMAP present",
        "SMPTE ST 429-9:2014",
        has_assetmap,
        if has_assetmap { "Found" } else { "Missing" },
    ));

    let has_volindex =
        opts.dcp_dir.join("VOLINDEX").exists() || opts.dcp_dir.join("VOLINDEX.xml").exists();
    report.structure_tests.push(make_test(
        "DCI-STRUCT-2",
        "VOLINDEX present",
        "SMPTE ST 429-9:2014",
        has_volindex,
        if has_volindex { "Found" } else { "Missing" },
    ));

    // Find PKL and CPL
    let (pkls, cpls, first_cpl) = find_xml_components(&opts.dcp_dir);

    report.structure_tests.push(make_test(
        "DCI-STRUCT-3",
        "PackingList (PKL) present",
        "SMPTE ST 429-8:2014",
        !pkls.is_empty(),
        if pkls.is_empty() {
            "No PKL found".into()
        } else {
            format!("Found {}", pkls.len())
        },
    ));

    report.structure_tests.push(make_test(
        "DCI-STRUCT-4",
        "CompositionPlaylist (CPL) present",
        "SMPTE ST 429-7:2006",
        !cpls.is_empty(),
        if cpls.is_empty() {
            "No CPL found".into()
        } else {
            format!("Found {}", cpls.len())
        },
    ));

    // MXF files
    let mxf_count = count_mxf_files(&opts.dcp_dir);
    report.structure_tests.push(make_test(
        "DCI-STRUCT-5",
        "MXF track files present",
        "SMPTE ST 429-3:2006",
        mxf_count > 0,
        format!("{mxf_count} MXF file(s)"),
    ));

    // --- CPL tests ---
    if let Some(ref cpl) = first_cpl {
        report.content_title.clone_from(&cpl.title);
        report.cpl_id.clone_from(&cpl.id);
        report.issue_date.clone_from(&cpl.issue_date);

        let valid_id = cpl.id.starts_with("urn:uuid:") && cpl.id.len() > 20;
        report.cpl_tests.push(make_test(
            "DCI-CPL-1",
            "CPL Id is valid URN UUID",
            "SMPTE ST 429-7:2006 §6.1",
            valid_id,
            &cpl.id,
        ));

        report.cpl_tests.push(make_test(
            "DCI-CPL-2",
            "ContentTitleText present",
            "SMPTE ST 429-7:2006 §6.2",
            !cpl.title.is_empty(),
            &cpl.title,
        ));

        report.cpl_tests.push(make_test(
            "DCI-CPL-3",
            "ContentKind present",
            "SMPTE ST 429-7:2006 §6.4",
            !cpl.content_kind.is_empty(),
            &cpl.content_kind,
        ));

        report.cpl_tests.push(make_test(
            "DCI-CPL-4",
            "At least one Reel present",
            "SMPTE ST 429-7:2006 §6.10",
            cpl.reel_count > 0,
            format!("{} reel(s)", cpl.reel_count),
        ));

        report.cpl_tests.push(make_test(
            "DCI-CPL-5",
            "IssueDate present",
            "SMPTE ST 429-7:2006 §6.3",
            !cpl.issue_date.is_empty(),
            &cpl.issue_date,
        ));
    }

    // --- Picture tests ---
    if opts.check_picture_profile {
        let empty_mxf = find_empty_mxf(&opts.dcp_dir);
        if empty_mxf.is_empty() {
            report.picture_tests.push(make_test(
                "DCI-PIC-1",
                "All MXF files have non-zero size",
                "SMPTE ST 429-3:2006",
                mxf_count > 0,
                "",
            ));
        } else {
            for name in &empty_mxf {
                report.picture_tests.push(make_test(
                    "DCI-PIC-1",
                    "MXF file has non-zero size",
                    "SMPTE ST 429-3:2006",
                    false,
                    format!("{name} is empty"),
                ));
            }
        }
    }

    // --- Audio tests ---
    if let Some(ref cpl) = first_cpl {
        report.audio_tests.push(make_test(
            "DCI-AUD-1",
            "Audio track referenced in CPL",
            "SMPTE ST 429-7:2006",
            cpl.has_audio,
            if cpl.has_audio {
                "MainSound present"
            } else {
                "No MainSound in any reel"
            },
        ));
    }

    // --- Security tests ---
    if opts.check_security {
        let has_encryption = cpls.iter().any(|p| {
            std::fs::read_to_string(p)
                .map(|c| c.contains("KeyId"))
                .unwrap_or(false)
        });
        report.security_tests.push(make_test(
            "DCI-SEC-1",
            "Encryption status detected",
            "SMPTE ST 429-6:2006",
            true,
            if has_encryption {
                "Encrypted (KeyId found)"
            } else {
                "Not encrypted"
            },
        ));
    }

    // Detect standard
    for cpl_path in &cpls {
        if let Ok(content) = std::fs::read_to_string(cpl_path) {
            if content.contains("smpte-ra.org") {
                report.detected_standard = Standard::Smpte;
            } else if content.contains("digicine.com") || content.contains("cinecert.com") {
                report.detected_standard = Standard::Interop;
            }
        }
    }

    // Summarize
    let all_tests = [
        &report.structure_tests,
        &report.cpl_tests,
        &report.picture_tests,
        &report.audio_tests,
        &report.security_tests,
    ];
    for tests in all_tests {
        for t in tests {
            report.total_tests += 1;
            if t.passed {
                report.tests_passed += 1;
            } else {
                report.tests_failed += 1;
            }
        }
    }
    report.conformant = report.tests_failed == 0;
    report
}

// --- Helpers ---

fn make_test(
    id: &str,
    desc: &str,
    spec: &str,
    passed: bool,
    detail: impl Into<String>,
) -> ConformanceTest {
    ConformanceTest {
        test_id: id.into(),
        description: desc.into(),
        spec_reference: spec.into(),
        passed,
        detail: detail.into(),
    }
}

fn today_iso() -> String {
    let now = time::OffsetDateTime::now_utc();
    format!(
        "{:04}-{:02}-{:02}",
        now.year(),
        now.month() as u8,
        now.day()
    )
}

struct CplInfo {
    id: String,
    title: String,
    content_kind: String,
    issue_date: String,
    reel_count: u32,
    has_audio: bool,
}

fn find_xml_components(dir: &Path) -> (Vec<PathBuf>, Vec<PathBuf>, Option<CplInfo>) {
    let mut pkls = Vec::new();
    let mut cpls = Vec::new();
    let mut first_cpl = None;

    let Ok(entries) = std::fs::read_dir(dir) else {
        return (pkls, cpls, first_cpl);
    };

    for entry in entries.flatten() {
        let path = entry.path();
        if !path.is_file() || path.extension().unwrap_or_default() != "xml" {
            continue;
        }
        let Ok(content) = std::fs::read_to_string(&path) else {
            continue;
        };

        if content.contains("PackingList") {
            pkls.push(path.clone());
        }
        if content.contains("<CompositionPlaylist") {
            cpls.push(path.clone());
            if first_cpl.is_none() {
                first_cpl = Some(parse_cpl_info(&content));
            }
        }
    }

    (pkls, cpls, first_cpl)
}

fn parse_cpl_info(content: &str) -> CplInfo {
    let id = extract_tag(content, "Id").unwrap_or_default();
    let title = extract_tag(content, "ContentTitleText").unwrap_or_default();
    let content_kind = extract_tag(content, "ContentKind").unwrap_or_default();
    let issue_date = extract_tag(content, "IssueDate").unwrap_or_default();

    let reel_re = regex_lite::Regex::new(r"<Reel>").unwrap();
    let reel_count = reel_re.find_iter(content).count() as u32;

    let has_audio = content.contains("<MainSound>");

    CplInfo {
        id,
        title,
        content_kind,
        issue_date,
        reel_count,
        has_audio,
    }
}

fn count_mxf_files(dir: &Path) -> u32 {
    std::fs::read_dir(dir)
        .map(|entries| {
            entries
                .flatten()
                .filter(|e| e.path().extension().is_some_and(|ext| ext == "mxf"))
                .count() as u32
        })
        .unwrap_or(0)
}

fn find_empty_mxf(dir: &Path) -> Vec<String> {
    std::fs::read_dir(dir)
        .map(|entries| {
            entries
                .flatten()
                .filter(|e| {
                    e.path().extension().is_some_and(|ext| ext == "mxf")
                        && e.metadata().map(|m| m.len() == 0).unwrap_or(false)
                })
                .map(|e| e.file_name().to_string_lossy().into_owned())
                .collect()
        })
        .unwrap_or_default()
}

fn extract_tag(xml: &str, tag: &str) -> Option<String> {
    let open = format!("<{tag}>");
    let close = format!("</{tag}>");
    let start = xml.find(&open)? + open.len();
    let end = xml[start..].find(&close)? + start;
    Some(xml[start..end].trim().to_string())
}
